runtime/vm/MHInterpreter.cpp

```diff
@@ -1353,6 +1366,127 @@ VM_MHInterpreter::getArgSlotsBeforePosition(j9object_t argumentTypes, U_32 argPo
 	return argumentSlots;
 }
 
+j9object_t
+VM_MHInterpreter::filterArgumentsWithCombiner(j9object_t methodHandle)
+{
+	/* Get filterHandle.type and filterHandle.type.argSlots */
+	/* [... filterHandle args] */
+	j9object_t filterHandleType = getMethodHandleMethodType(methodHandle);
+	j9object_t argumentTypes = getMethodTypeArguments(filterHandleType);
+	U_32 filterArgSlots = getMethodTypeArgSlots(filterHandleType);
+	j9object_t argumentIndices = J9VMJAVALANGINVOKEFILTERARGUMENTSWITHCOMBINERHANDLE_ARGUMENTINDICES(_currentThread, methodHandle);
+	U_32 argumentIndicesCount = J9INDEXABLEOBJECT_SIZE(_currentThread, argumentIndices);
+	UDATA *spFirstFilterArgSlot = _currentThread->sp + filterArgSlots;
+
+	/* [... filterHandle args descriptionBytes MethodTypeFrame] */
+	UDATA *spPriorToFrameBuild = _currentThread->sp;
+	(void)buildMethodTypeFrame(_currentThread, filterHandleType);
+
+	/* [... filterHandle args descriptionBytes MethodTypeFrame filterHandle args] */
+	_currentThread->sp -= (filterArgSlots  + 1);
+	memcpy(_currentThread->sp, spPriorToFrameBuild, sizeof(UDATA) * (filterArgSlots+1));
+
+	/* [... filterHandle args descriptionBytes MethodTypeFrame filterHandle PlaceHolderFrame filterHandle args] */
+	insertPlaceHolderFrame(filterArgSlots, methodHandle, J9VMJAVALANGINVOKEMETHODHANDLE_FILTERARGUMENTSWITHCOMBINERPLACEHOLDER_METHOD(_vm));
+
+	/* Get combinerHandle */
+	j9object_t combinerHandle = getCombinerHandleForFilter(methodHandle);
+
+	/* Replace filterHandle with combinerHandle and adjust sp according to combinerHandle.type.argSlots */
+	/* [... filterHandle args descriptionBytes MethodTypeFrame filterHandle PlaceHolderFrame combinerHandle filterArgs] */
+	((j9object_t *)_currentThread->sp)[filterArgSlots] = combinerHandle;
+	UDATA *spCombinerSlot = _currentThread->sp + filterArgSlots;
+
+	/* Copy all arguments specified by argumentIndices from filterHandle to the stack slots used by combinerHandler */
+	U_32 arrayIndex = 0;
+	for (arrayIndex = 0; arrayIndex < argumentIndicesCount; arrayIndex++) {
+		U_32 argumentTypeIndex = (U_32)J9JAVAARRAYOFINT_LOAD(_currentThread, argumentIndices, arrayIndex);
+
+		/* Determine the slot index of each argument specified by the index in argumentIndices */
+		U_32 argumentTypeSlots = getArgSlotsBeforePosition(argumentTypes, argumentTypeIndex);
```

Comment:
This is `N*M` in terms of the number of arguments.  How many arguments are commonly used here?  

If the common case from the class libraries is 2 or 3 arguments, this isn't so bad.  If it's more than that, we'll want to do something better here.

Reply:
I'm seeing argument lengths of only 2 & 3

Reply:
Ok.  Let's leave it for now as the JIT implementation won't have this problem.  If it shows up in profiles, we can look for other ways to optimize it